### DISCUSSION TITLE
Running CUDA Quantum NVIDIA GPU Simulations for multi-node and multi-GPU scenarios

### DISCUSSION THREAD

**Discussion by ppats15** [NVIDIA/cuda-quantum#484] [Q&A]
Are there any instructions on how to run multi-node and multi-GPU scenarios? Which target should be selected? 
From the latest release, I see: 
nvidia-mqpu-fp64
nvidia-mqpu 
nvidia 
What is the difference between these options, and which one should be selected for the multi-node scenario?
Appreciate any help?

**Comment by Advait251206** [comment]
This is a very common confusion in CUDA Quantum because the naming looks like it maps cleanly to “multi-node”, but it actually doesn’t.

Let’s separate the concepts clearly.

---

# 1. What these targets actually mean

## ✔ `nvidia`
This is the **standard single-node GPU backend**.

- runs on one machine
- may use 1 or multiple GPUs (depending on configuration)
- no distributed communication layer

👉 Think: *“normal CUDA execution”*

---

## ✔ `nvidia-mqpu`
This means:

> multi-GPU (within a single node)

- multiple GPUs on the same host
- uses NCCL / CUDA device coordination
- shared memory / PCIe / NVLink

👉 NOT multi-node

---

## ✔ `nvidia-mqpu-fp64`
Same as `nvidia-mqpu`, but:

- uses **double precision (FP64) state representation**
- higher accuracy
- slower performance
- more memory usage

---

# 2. Key misunderstanding: MQPU ≠ multi-node

This is the most important point:

> ❌ `mqpu` does NOT mean multi-node  
> ✔ it means multi-GPU *within one machine*

So:

| Target | Multi-GPU | Multi-node |
|--------|----------|-----------|
| nvidia | optional (single GPU default) | ❌ |
| nvidia-mqpu | ✔ | ❌ |
| nvidia-mqpu-fp64 | ✔ | ❌ |

---

# 3. So what is used for multi-node?

CUDA Quantum does NOT expose a separate simple “multi-node target” like:

```
nvidia-mqpu-multinode
```

Instead, multi-node support depends on:

## ✔ MPI / distributed launcher
You must use:

- `mpirun`
- or Slurm (`srun`)
- or cluster orchestration

and run the **same `nvidia-mqpu` or SVS backend across nodes**

---

## ✔ statevector backend case

For statevector simulations:

- distribution happens via **SVS (State Vector Simulator) backend**
- not via `nvidia-mqpu`

So multi-node is usually:

> MPI + SVS backend, not MQPU

---

# 4. Correct mental model

Think of CUDA Quantum targets like this:

```
nvidia
 ├── single GPU / basic execution

nvidia-mqpu
 ├── multi-GPU (single node only)

svs (statevector simulator)
 ├── can scale multi-GPU and multi-node via MPI

tensor network backends
 ├── may support distributed execution depending on build
``` id="mqpu_model"
```

---

# 5. Why MQPU is not multi-node

Because MQPU assumes:

- low-latency GPU communication (NVLink / PCIe)
- shared host memory domain
- no network serialization layer

Multi-node requires:

- MPI message passing
- distributed state partitioning
- network-aware synchronization

👉 That is a completely different runtime layer

---

# 6. Which should you choose?

## ✔ If you are on ONE machine with multiple GPUs:

```bash
cudaq.set_target("nvidia-mqpu")
```

---

## ✔ If you want multi-node scaling:

You should NOT switch to MQPU.

Instead:

- use MPI launch (`mpirun`, `srun`)
- use SVS backend (if applicable)
- ensure distributed runtime is enabled

---

# 7. Common mistake

Many users try:

```
nvidia-mqpu + multiple nodes
```

and expect:

> automatic cluster scaling

But what actually happens:

- each node runs independent simulation OR
- it silently falls back to single-node behavior

---

# 8. Direct answer

- ❌ `nvidia-mqpu` is NOT for multi-node
- ✔ it is for multi-GPU single-node execution
- ✔ `nvidia` is single-GPU/single-node baseline
- ✔ multi-node requires MPI + SVS or distributed backend, not MQPU

---

# If you want

I can give you:

- a working **multi-node CUDA-Q SVS launch script**
- or a diagram of how MQPU vs SVS actually distribute statevectors
- or a checklist to verify whether your job is actually multi-node or silently single-node

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206

**Comment by rogerawong** [comment]
The reply above states that **mqpu** can't span nodes. That's incorrect. It can by way of the **MPI distribution mode**.

This question is three years old, and the developer environment has changed quite a bit since then. The OP lists **nvidia-mqpu-fp64**, **nvidia-mqpu**, **nvidia**, and those are all the pre-0.8 layout. Those separate targets were collapsed into one **nvidia** target with composable options:

> "Deprecated since version 0.8: The **nvidia-mqpu** and **nvidia-mqpu-fp64** targets, which are equivalent to the multi-QPU options **mqpu,fp32** and **mqpu,fp64**, respectively, of the **nvidia** target, are deprecated and will be removed in a future release."

So **nvidia-mqpu** is now **option="mqpu"**, and **nvidia-mqpu-fp64** is now **option="mqpu,fp64**"

An important option missing from the OP's list is **mgpu**. Both mgpu and mqpu can run across nodes. But it's not about how many machines you have, it's about choosing what work you're distributing. Which you choose depends on the problem you're solving.

- **mgpu**. Splits one state vector across GPUs and nodes. Use when the state vector won't fit in a single GPU's memory.
- **mqpu**. Gives you one simulated QPU per GPU, run asynchronously. Use when you have many independent circuits to run in parallel. 

Whether you're limited to a single node depends on the distribution mode you pick, not the target.
>Thread: distribute the expectation value computations among available GPUs via standard C++ threads (each thread handles one GPU).

>For instance, if all GPUs are available on a single node, thread-based parallel distribution (cudaq::parallel::thread in C++ or cudaq.parallel.thread in Python, as shown in the above example) is sufficient. On the other hand, if one wants to distribute the tasks across GPUs on multiple nodes, e.g., on a compute cluster, MPI distribution mode should be used.

Thread mode already uses every GPU available on a machine. To cross the node boundary, you use MPI mode:

One big circuit across nodes:
```bash
mpiexec -np 2 python3 program.py --target nvidia --target-option fp64,mgpu
```

Lots of circuits across nodes:

```python
cudaq.mpi.initialize()
cudaq.set_target("nvidia", option="mqpu")
cudaq.observe(kernel, hamiltonian, angle, execution=cudaq.parallel.mpi)
cudaq.mpi.finalize()
```
Then run it with:
```bash
mpiexec -np <N> python3 file.py
```

Notes:
- **ranks_per_qpu** backs each virtual QPU with more than one rank and GPU, so a single **observe** call can simulate circuits that exceed one GPU’s memory while you still run many QPUs across the cluster. That’s the case where the choice stops being either/or.
- **remote-mqpu** is gone. Direct MPI distribution replaced it, so you have to decide how to distribute ranks (workers) through scheduler flags, like SLURM’s **--ntasks-per-node**.
- if you installed CUDA-Q with **pip**, the MPI dependencies install separately.

Sources: [State vector simulators, multi-GPU multi-node](https://nvidia.github.io/cuda-quantum/latest/using/backends/sims/svsims.html#multi-gpu-multi-node) and [Multiple QPUs, parallel distribution mode](https://nvidia.github.io/cuda-quantum/latest/using/backends/sims/mqpusims.html#parallel-distribution-mode)
